File: src/solver/definitions.lua

local helperFuns = require "helpers"

local definitions = {}
definitions.rowNames = {'A','B','C','D','E','F','G','H','I'}
definitions.colNames = {'1','2','3','4','5','6','7','8','9'}
definitions.allKeys  = helperFuns.cross(definitions.rowNames, definitions.colNames)

definitions.cellRows    = {
                        [1] = 'ABC',
                        [2] = 'DEF',
                        [3] = 'GHI'}
definitions.cellColumns = {
                        [1] = '123',
                        [2] = '456',
                        [3] = '789'}

definitions.difficultEnum = {
    [1] = 'TRIVIAL',
    [2] = 'EASY',
    [3] = 'SO SO',
    [4] = 'HARD',
    [5] = 'VERY HARD',
    [6] = 'EVIL',
    [7] = 'DASTURDLY EVIL'}

return definitions

File: src/solver/definitions.jl
module Definitions

export rowNames, columnNames, allKeys, neighbors, cellRows, cellColumns, puzzle0

const rowNames::String             = "ABCDEFGHI"
const columnNames::String         = "123456789"
const allKeys::Vector{String}     = unique([string(row, col) for row in rowNames for col in columnNames])
const squares::Vector{String}     = allKeys
cellRows    = ["ABC", "DEF", "GHI"]
cellColumns = ["123", "456", "789"]


function _neighborsOf(sq::String)
    local _colCell = cellColumns[(first(findfirst(sq[2], columnNames))%3)+1]
    local _rowCell = cellRows[(first(findfirst(sq[1], rowNames))%3)+1]

    local _neighborCells = [string(row, col) for row in _rowCell for col in _colCell]
    local _neighborrowNames = [string(sq[1], col) for col in columnNames]
    local _neighborCols = [string(row, sq[2]) for row in rowNames]

    return filter!(!=(sq),sort(unique(vcat(_neighborCells, _neighborrowNames, _neighborCols))))

end

# Get all the neighbors for every square
const neighbors = Dict{String,Vector{String}}(sq => _neighborsOf(sq) for sq in squares)

# Default empty puzzle
puzzle0 = Dict{String,String}(sq => "123456789" for sq in squares)

end